File: src/main/kotlin/com/ribafish/slacksummarizer/services/SlackService.kt
package com.ribafish.slacksummarizer.services

import com.ribafish.slacksummarizer.config.SlackConfig
import com.ribafish.slacksummarizer.models.SlackMessage
import com.ribafish.slacksummarizer.models.SlackThread
import com.slack.api.Slack
import com.slack.api.methods.MethodsClient
import com.slack.api.methods.request.conversations.ConversationsRepliesRequest
import mu.KotlinLogging

private val logger = KotlinLogging.logger {}

class SlackService(private val config: SlackConfig) {
    private val slack: Slack = Slack.getInstance()
    private val client: MethodsClient = slack.methods(config.botToken)

    suspend fun fetchThread(channelId: String, threadTs: String): SlackThread {
        logger.debug { "Fetching thread $threadTs from channel $channelId" }

        // Try to join the channel if it's public (will fail silently for private channels)
        try {
            val joinResponse = client.conversationsJoin { req ->
                req.channel(channelId)
            }
            if (joinResponse.isOk) {
                logger.debug { "Joined channel $channelId" }
            } else if (joinResponse.error != "already_in_channel") {
                logger.debug { "Could not join channel $channelId: ${joinResponse.error}" }
            }
        } catch (e: Exception) {
            logger.debug(e) { "Exception joining channel $channelId: ${e.message}" }
        }

        // Get channel info
        val channelInfo = client.conversationsInfo { req ->
            req.channel(channelId)
        }

        val channelName = if (channelInfo.isOk) {
            channelInfo.channel.name ?: channelId
        } else {
            logger.warn { "Failed to get channel info: ${channelInfo.error}" }
            channelId
        }

        // Get thread messages
        val repliesRequest = ConversationsRepliesRequest.builder()
            .channel(channelId)
            .ts(threadTs)
            .build()

        val repliesResponse = client.conversationsReplies(repliesRequest)

        if (!repliesResponse.isOk) {
            logger.error { "Failed to fetch thread: ${repliesResponse.error}" }
            return SlackThread(channelId, channelName, threadTs, emptyList())
        }

        val messages = repliesResponse.messages.map { message ->
            val userId = message.user ?: "unknown"
            val username = getUserName(userId)

            SlackMessage(
                user = userId,
                username = username,
                text = message.text ?: "",
                timestamp = message.ts
            )
        }

        logger.debug { "Fetched ${messages.size} messages from thread" }

        return SlackThread(
            channelId = channelId,
            channelName = channelName,
            threadTs = threadTs,
            messages = messages
        )
    }

    private fun getUserName(userId: String): String {
        return try {
            val userInfo = client.usersInfo { req ->
                req.user(userId)
            }

            if (userInfo.isOk) {
                userInfo.user.realName ?: userInfo.user.name ?: userId
            } else {
                logger.warn { "Failed to get user info for $userId: ${userInfo.error}" }
                userId
            }
        } catch (e: Exception) {
            logger.warn(e) { "Error fetching user info for $userId" }
            userId
        }
    }
}

File: summarizer-python/services/slack_service.py
"""Slack API integration service."""

import logging
from typing import List

from slack_sdk import WebClient
from slack_sdk.errors import SlackApiError

from ..config import SlackConfig
from ..models import SlackMessage, SlackThread

logger = logging.getLogger(__name__)


class SlackService:
    """Service for interacting with Slack API."""

    def __init__(self, config: SlackConfig):
        """Initialize Slack service with configuration."""
        self.config = config
        self.client = WebClient(token=config.bot_token)

    def post_message(self, channel_id: str, text: str, thread_ts: str = None) -> None:
        """Post a message to a Slack channel or thread.

        Args:
            channel_id: The channel ID
            text: The message text
            thread_ts: Optional thread timestamp to reply in thread
        """
        try:
            kwargs = {
                "channel": channel_id,
                "text": text
            }
            if thread_ts:
                kwargs["thread_ts"] = thread_ts

            response = self.client.chat_postMessage(**kwargs)
            if response["ok"]:
                logger.debug(f"Posted message to channel {channel_id}")
            else:
                logger.error(f"Failed to post message: {response.get('error')}")
        except SlackApiError as e:
            logger.error(f"Failed to post message: {e.response['error']}")

    def fetch_thread(self, channel_id: str, thread_ts: str) -> SlackThread:
        """Fetch a thread from Slack.

        Args:
            channel_id: The channel ID
            thread_ts: The thread timestamp

        Returns:
            SlackThread containing all messages
        """
        logger.debug(f"Fetching thread {thread_ts} from channel {channel_id}")

        # Try to join the channel if it's public
        try:
            join_response = self.client.conversations_join(channel=channel_id)
            if join_response["ok"]:
                logger.debug(f"Joined channel {channel_id}")
        except SlackApiError as e:
            if e.response["error"] != "already_in_channel":
                logger.debug(f"Could not join channel {channel_id}: {e.response['error']}")

        # Get workspace/team info
        workspace_id = None
        try:
            team_info = self.client.team_info()
            workspace_id = team_info["team"]["id"]
        except Exception as e:
            logger.debug(f"Could not get team info: {e}")

        # Get channel info
        channel_name = channel_id
        try:
            channel_info = self.client.conversations_info(channel=channel_id)
            if channel_info["ok"]:
                channel_name = channel_info["channel"]["name"]
        except SlackApiError as e:
            logger.warning(f"Failed to get channel info: {e.response['error']}")

        # Get thread messages
        try:
            replies_response = self.client.conversations_replies(
                channel=channel_id,
                ts=thread_ts
            )

            if not replies_response["ok"]:
                logger.error(f"Failed to fetch thread: {replies_response.get('error', 'unknown error')}")
                return SlackThread(
                    channel_id=channel_id,
                    channel_name=channel_name,
                    thread_ts=thread_ts,
                    messages=[],
                    workspace_id=workspace_id
                )

            messages = []
            for message in replies_response["messages"]:
                messages.append(SlackMessage(
                    user=message.get("user", "unknown"),
                    username="",  # Not needed for knowledge base
                    text=message.get("text", ""),
                    timestamp=message["ts"]
                ))

            logger.debug(f"Fetched {len(messages)} messages from thread")

            return SlackThread(
                channel_id=channel_id,
                channel_name=channel_name,
                thread_ts=thread_ts,
                messages=messages,
                workspace_id=workspace_id
            )

        except SlackApiError as e:
            logger.error(f"Failed to fetch thread: {e.response['error']}")
            return SlackThread(
                channel_id=channel_id,
                channel_name=channel_name,
                thread_ts=thread_ts,
                messages=[],
                workspace_id=workspace_id
            )
